File: packages/desktop-helpers/macos/Sources/CorivoCaptureHelper/OCR/OcrHandlers.swift
//
//  OcrHandlers.swift
//  RPC handler for `ocr.run`.
//

import Foundation

enum OcrHandlers {

    private struct Payload: Decodable {
        var imagePath: String
        var languages: [String]?
        var useLanguageCorrection: Bool?

        enum CodingKeys: String, CodingKey {
            case imagePath = "image_path"
            case languages
            case useLanguageCorrection = "use_language_correction"
        }
    }

    static func run(_ request: RequestMessage) async throws {
        let payload = try decodePayload(request.payload, as: Payload.self)
        let req = VisionOCR.Request(
            imagePath: payload.imagePath,
            languages: payload.languages ?? ["zh-Hans", "zh-Hant", "en-US"],
            useLanguageCorrection: payload.useLanguageCorrection ?? false
        )
        let result = try await Task.detached(priority: .userInitiated) {
            try VisionOCR.run(req)
        }.value
        let json: [String: Any] = [
            "text": result.text,
            "elapsed_ms": result.elapsedMs,
        ]
        try await EventEmitter.shared.send(
            ResponseMessage(id: request.id, result: AnyCodable(json))
        )
    }

    private static func decodePayload<T: Decodable>(
        _ payload: AnyCodable?, as _: T.Type
    ) throws -> T {
        guard let p = payload else {
            throw HelperError.invalidRequest("missing payload")
        }
        let data = try JSONCodec.encoder.encode(p)
        return try JSONCodec.decoder.decode(T.self, from: data)
    }
}

File: packages/desktop-helpers/windows/src/ocr/ocr_handlers.cpp
// ocr_handlers.cpp

#include "ocr_handlers.hpp"

#include <stdexcept>

#include "windows_ocr.hpp"
#include "../ipc/event_emitter.hpp"

namespace corivo::ocr {

void handle_run(const corivo::proto::Request& request) {
    using namespace corivo::proto;

    auto& payload = request.payload;
    if (!payload.is_object()) {
        // Qualify explicitly: an unqualified `Response` resolves to
        // `corivo::ocr::Response` (the OCR result type defined in
        // windows_ocr.hpp) because we're inside `namespace corivo::ocr`,
        // and the `using namespace corivo::proto;` above doesn't beat
        // closer scope. We want the wire response here.
        corivo::proto::Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{
            error_codes::INVALID_REQUEST,
            "ocr.run: payload must be an object",
            nullptr,
        };
        corivo::ipc::EventEmitter::instance().send(resp);
        return;
    }

    Request req;
    if (auto it = payload.find("image_path");
        it != payload.end() && it->is_string()) {
        req.image_path = it->get<std::string>();
    }
    if (req.image_path.empty()) {
        // Qualify explicitly: an unqualified `Response` resolves to
        // `corivo::ocr::Response` (the OCR result type defined in
        // windows_ocr.hpp) because we're inside `namespace corivo::ocr`,
        // and the `using namespace corivo::proto;` above doesn't beat
        // closer scope. We want the wire response here.
        corivo::proto::Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{
            error_codes::INVALID_REQUEST,
            "ocr.run: missing 'image_path'",
            nullptr,
        };
        corivo::ipc::EventEmitter::instance().send(resp);
        return;
    }
    if (auto it = payload.find("languages");
        it != payload.end() && it->is_array()) {
        for (auto const& v : *it) {
            if (v.is_string()) req.languages.push_back(v.get<std::string>());
        }
    }
    if (req.languages.empty()) {
        req.languages = {"zh-Hans", "zh-Hant", "en-US"};
    }
    if (auto it = payload.find("use_language_correction");
        it != payload.end() && it->is_boolean()) {
        req.use_language_correction = it->get<bool>();
    }

    try {
        auto out = run(req);
        nlohmann::json result = {
            {"text", out.text},
            {"elapsed_ms", out.elapsed_ms},
        };
        // Qualify explicitly: an unqualified `Response` resolves to
        // `corivo::ocr::Response` (the OCR result type defined in
        // windows_ocr.hpp) because we're inside `namespace corivo::ocr`,
        // and the `using namespace corivo::proto;` above doesn't beat
        // closer scope. We want the wire response here.
        corivo::proto::Response resp;
        resp.id = request.id;
        resp.ok = true;
        resp.result = result;
        corivo::ipc::EventEmitter::instance().send(resp);
    } catch (const std::exception& e) {
        // Qualify explicitly: an unqualified `Response` resolves to
        // `corivo::ocr::Response` (the OCR result type defined in
        // windows_ocr.hpp) because we're inside `namespace corivo::ocr`,
        // and the `using namespace corivo::proto;` above doesn't beat
        // closer scope. We want the wire response here.
        corivo::proto::Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{error_codes::OS_ERROR, e.what(), nullptr};
        corivo::ipc::EventEmitter::instance().send(resp);
    }
}

} // namespace corivo::ocr
